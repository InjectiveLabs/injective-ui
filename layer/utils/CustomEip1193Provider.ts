import {
  loadEthersSigningType,
  loadEthersBaseWalletType,
  loadSigUtilSignedTypedData
} from './../utils/lib'
import type { BaseWallet, Eip1193Provider } from 'ethers'

const SEPOLIA_CHAIN_ID = 11155111

const methodMap: Record<
  string,
  (
    params: any,
    pk: string,
    baseWallet: BaseWallet,
    provider: CustomEip1193Provider
  ) => Promise<any>
> = {
  eth_requestAccounts: async (_method, _params, _baseWallet) => {
    return [_baseWallet.address]
  },
  eth_signTypedData_v4: async (params, pk) => {
    const signTypedData = await loadSigUtilSignedTypedData()
    const _pk = pk.slice(2)

    const [_address, typedData] = params

    const typedDataObj = JSON.parse(typedData) as any

    const { domain, types, message, primaryType } = typedDataObj

    const signature = signTypedData({
      data: { domain, message, primaryType, types },
      privateKey: Buffer.from(_pk, 'hex'),
      version: 'v4' as any
    })

    return signature
  },
  eth_chainId: async (_method, _params, _baseWallet, _provider) => {
    return '0x' + _provider.chainId.toString(16)
  },
  wallet_switchEthereumChain: async (
    _method,
    _params,
    _baseWallet,
    _provider
  ) => {
    const [chainId] = _params

    const chainIdObj = JSON.parse(chainId) as any

    const chainIdInt = parseInt(chainIdObj.chainId, 16)

    _provider.chainId = chainIdInt
  }
}

export class CustomEip1193Provider implements Eip1193Provider {
  isMetaMask: boolean
  chainId: number

  private baseWallet: undefined | BaseWallet
  private pk: string | undefined

  constructor(chainId: number = SEPOLIA_CHAIN_ID) {
    this.isMetaMask = true

    this.chainId = chainId

    const windowOrGlobal =
      typeof window !== 'undefined' ? window : (globalThis as any)

    windowOrGlobal.setPrivateKey = this.setPrivateKey.bind(this)
  }

  async request({ method, params }: { params?: any; method: string }) {
    if (!this.pk || !this.baseWallet) {
      throw new Error('Private key not set')
    }

    if (method in methodMap) {
      return methodMap[method](params, this.pk, this.baseWallet, this)
    }

    throw new Error(`Method ${method} not found`)
  }

  async setPrivateKey(pk: string) {
    const SigningKey = await loadEthersSigningType()
    const BaseWallet = await loadEthersBaseWalletType()

    this.pk = pk

    const signingKey = new SigningKey(pk)
    this.baseWallet = new BaseWallet(signingKey)
  }
}
