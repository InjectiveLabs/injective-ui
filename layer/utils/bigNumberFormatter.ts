import {
  BigNumber,
  BigNumberInBase,
  getExactDecimalsFromNumber
} from '@injectivelabs/utils'
import {
  ZERO_IN_BASE,
  INJ_REQUIRED_FOR_GAS,
  DEFAULT_ROUNDING_MODE,
  DEFAULT_BIG_NUMBER_DECIMAL_PLACES
} from './constant'

export const DEFAULT_MINIMAL_DISPLAY_DECIMAL_PLACES = 4
export const DEFAULT_MAX_MINIMAL_DISPLAY_DECIMAL_PLACES = 12

export const getFormattedZeroValue = (decimalPlaces: number): string => {
  if (decimalPlaces > 0) {
    return `0.${'0'.repeat(decimalPlaces)}`
  }
  return '0.0'
}

export const unAbbreviateNumber = (
  value: string
): BigNumberInBase | undefined => {
  const units = {
    K: Number(`1${'0'.repeat(3)}`),
    M: Number(`1${'0'.repeat(6)}`),
    B: Number(`1${'0'.repeat(9)}`),
    T: Number(`1${'0'.repeat(12)}`)
  } as Record<string, number>

  const unit = value.at(-1)

  if (!unit || !units[unit]) {
    return
  }

  const formattedValue = value.replaceAll(',', '').slice(0, -1)

  return new BigNumberInBase(formattedValue).multipliedBy(units[unit])
}

export const abbreviateNumber = (number: number): string => {
  const abbreviatedValue = new Intl.NumberFormat('en-US', {
    notation: 'compact',
    compactDisplay: 'short'
  }).format(number)

  const abbreviatedValueMatchesInput = new BigNumberInBase(number).eq(
    unAbbreviateNumber(abbreviatedValue) || '0'
  )

  return abbreviatedValueMatchesInput
    ? abbreviatedValue
    : `â‰ˆ${abbreviatedValue}`
}

export const getNumberMinimalDecimals = (
  value: BigNumberInBase,
  defaultMinimalDecimalPlaces?: number,
  displayAbsoluteDecimalPlace?: boolean
) => {
  const valueExactDecimals = getExactDecimalsFromNumber(value.toFixed())
  const minimalDecimalPlaces =
    defaultMinimalDecimalPlaces ?? DEFAULT_MINIMAL_DISPLAY_DECIMAL_PLACES
  const minNumberFromDefaultMinDecimals = new BigNumberInBase(1).shiftedBy(
    -minimalDecimalPlaces
  )

  if (value.eq(0)) {
    return {
      minimalDecimalPlaces: 2,
      minimalDisplayAmount: ZERO_IN_BASE
    }
  }

  /**
   * The number is less then the range of minimal decimals,
   * for example if minimalDecimalPlaces = 4, the number is
   * smaller than 0.0001
   * show exact decimal places the based on the value provided up to 12 decimal places
   * i.e a value with 0.000001010102 will be displayed as < 0.000001
   */
  if (
    value.lte(minNumberFromDefaultMinDecimals) &&
    displayAbsoluteDecimalPlace
  ) {
    const actualMinimalDecimalPlaces =
      valueExactDecimals > DEFAULT_MAX_MINIMAL_DISPLAY_DECIMAL_PLACES
        ? DEFAULT_MAX_MINIMAL_DISPLAY_DECIMAL_PLACES
        : valueExactDecimals
    const minimalDisplayAmount = new BigNumber(1).shiftedBy(
      -actualMinimalDecimalPlaces
    )

    /**
     * Go up to 12 decimals if needed, which is a maximum we can support
     * at this point
     */
    return {
      minimalDisplayAmount,
      minimalDecimalPlaces: actualMinimalDecimalPlaces
    }
  }

  /*
   * display the value based on the minimalDecimalPlaces value provided
   * which is default to 4, if not provided
   * for example minimalDecimalPlaces: 3 will show <0.01 if value is less then 0.01
   */
  return {
    minimalDecimalPlaces,
    minimalDisplayAmount: new BigNumber(1).shiftedBy(-minimalDecimalPlaces)
  }
}

export const convertToBigNumber = (
  value: string | number | BigNumberInBase
): BigNumberInBase => {
  if (!value) {
    return new BigNumberInBase(0)
  }

  if (BigNumber.isBigNumber(value)) {
    return value as BigNumberInBase
  }

  return new BigNumberInBase(value as string)
}

export const convertValueToFixed = ({
  options,
  roundingMode,
  decimalPlaces,
  valueToBigNumber
}: {
  options: {
    shouldTruncate?: boolean
    abbreviationFloor?: number
  }
  roundingMode: BigNumber.RoundingMode
  decimalPlaces: number
  valueToBigNumber: BigNumberInBase
}): string => {
  if (valueToBigNumber.isNaN() || valueToBigNumber.isZero()) {
    return options.shouldTruncate ? '0' : getFormattedZeroValue(decimalPlaces)
  }

  if (
    options.abbreviationFloor &&
    valueToBigNumber.gte(options.abbreviationFloor)
  ) {
    return abbreviateNumber(valueToBigNumber.toNumber())
  }

  const roundedValue = valueToBigNumber.toFixed(decimalPlaces, roundingMode)

  if (options.shouldTruncate) {
    return new BigNumberInBase(roundedValue).toFixed()
  }

  return roundedValue
}

export const convertValueToString = ({
  options,
  roundingMode,
  decimalPlaces,
  valueToBigNumber
}: {
  options: {
    shouldTruncate?: boolean
    abbreviationFloor?: number
    minimalDecimalPlaces?: number
    displayAbsoluteDecimalPlace?: boolean
  }
  roundingMode: BigNumber.RoundingMode
  decimalPlaces: number
  valueToBigNumber: BigNumberInBase
}): string => {
  if (valueToBigNumber.isNaN() || valueToBigNumber.isZero()) {
    return options.shouldTruncate ? '0' : getFormattedZeroValue(decimalPlaces)
  }

  if (
    options.abbreviationFloor &&
    valueToBigNumber.gte(options.abbreviationFloor)
  ) {
    return abbreviateNumber(valueToBigNumber.toNumber())
  }

  const { minimalDecimalPlaces, minimalDisplayAmount } =
    getNumberMinimalDecimals(
      valueToBigNumber,
      options.minimalDecimalPlaces,
      options.displayAbsoluteDecimalPlace
    )

  const roundedValue = valueToBigNumber.toFixed(decimalPlaces, roundingMode)

  if (options.shouldTruncate) {
    return new BigNumberInBase(roundedValue).toFormat()
  }

  if (valueToBigNumber.abs().lt(minimalDisplayAmount)) {
    return `< ${minimalDisplayAmount.toFormat(minimalDecimalPlaces)}`
  }

  return new BigNumberInBase(roundedValue).toFormat(decimalPlaces)
}

export const computeValueWithGasBuffer = (
  valueToBigNumber: BigNumberInBase,
  injFee: number
): BigNumberInBase => {
  if (valueToBigNumber.isNaN() || valueToBigNumber.lte(injFee)) {
    return new BigNumberInBase(0)
  }

  return valueToBigNumber.minus(injFee)
}

export const computeValueWithGasBufferToFixed = ({
  roundingMode,
  decimalPlaces,
  valueWithGasBuffer
}: {
  roundingMode: BigNumber.RoundingMode
  decimalPlaces: number
  valueWithGasBuffer: BigNumberInBase
}): string => {
  if (valueWithGasBuffer.isNaN() || valueWithGasBuffer.isZero()) {
    return getFormattedZeroValue(decimalPlaces)
  }

  return valueWithGasBuffer.toFixed(decimalPlaces, roundingMode)
}

export const computeValueWithGasBufferToString = ({
  options,
  roundingMode,
  decimalPlaces,
  valueWithGasBuffer
}: {
  options: {
    abbreviationFloor?: number
    minimalDecimalPlaces?: number
    displayAbsoluteDecimalPlace?: boolean
  }
  roundingMode: BigNumber.RoundingMode
  decimalPlaces: number
  valueWithGasBuffer: BigNumberInBase
}): string => {
  if (valueWithGasBuffer.isNaN() || valueWithGasBuffer.isZero()) {
    return getFormattedZeroValue(decimalPlaces)
  }

  if (
    options.abbreviationFloor &&
    valueWithGasBuffer.gte(options.abbreviationFloor)
  ) {
    return abbreviateNumber(valueWithGasBuffer.toNumber())
  }

  const { minimalDecimalPlaces, minimalDisplayAmount } =
    getNumberMinimalDecimals(
      valueWithGasBuffer,
      options.minimalDecimalPlaces,
      options.displayAbsoluteDecimalPlace
    )

  if (valueWithGasBuffer.abs().lt(minimalDisplayAmount)) {
    return `< ${minimalDisplayAmount.toFormat(minimalDecimalPlaces)}`
  }

  return valueWithGasBuffer.toFormat(decimalPlaces, roundingMode)
}

export const sharedBigNumberFormatter = (
  value: string | number | BigNumberInBase,
  options: {
    decimalPlaces?: number
    minimalDecimalPlaces?: number
    abbreviationFloor?: number // Whether to abbreviate numbers, e.g., 1,234,455 => 1M
    injFee?: number
    roundingMode?: BigNumber.RoundingMode
    displayAbsoluteDecimalPlace?: boolean // Explained above
    shouldTruncate?: boolean
  } = {}
) => {
  const roundingMode = options.roundingMode || DEFAULT_ROUNDING_MODE
  const decimalPlaces =
    options.decimalPlaces ?? DEFAULT_BIG_NUMBER_DECIMAL_PLACES
  const injFee = options.injFee ?? INJ_REQUIRED_FOR_GAS

  const valueToBigNumber = convertToBigNumber(value)

  const valueToFixed = convertValueToFixed({
    options,
    roundingMode,
    decimalPlaces,
    valueToBigNumber
  })

  const valueToString = convertValueToString({
    options,
    roundingMode,
    decimalPlaces,
    valueToBigNumber
  })

  const valueWithGasBuffer = computeValueWithGasBuffer(valueToBigNumber, injFee)

  const valueWithGasBufferToFixed = computeValueWithGasBufferToFixed({
    roundingMode,
    decimalPlaces,
    valueWithGasBuffer
  })

  const valueWithGasBufferToString = computeValueWithGasBufferToString({
    options,
    roundingMode,
    decimalPlaces,
    valueWithGasBuffer
  })

  return {
    valueToFixed,
    valueToString,
    valueToBigNumber,
    valueWithGasBuffer,
    valueWithGasBufferToFixed,
    valueWithGasBufferToString
  }
}
