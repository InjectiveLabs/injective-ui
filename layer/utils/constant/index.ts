import {
  BigNumber,
  BigNumberInWei,
  BigNumberInBase
} from '@injectivelabs/utils'

export * from './setup'

export const INJ_LOGO_URL =
  'https://imagedelivery.net/DYKOWp0iCc0sIkF-2e4dNw/c7b97294-890e-459c-4673-1ca1dea64f00/public'
export const INJ_LOGO_DARK_URL =
  'https://imagedelivery.net/DYKOWp0iCc0sIkF-2e4dNw/8493baf6-e4d3-4ac6-5a3b-afe15b47ae00/public'
export const USDT_LOGO_URL =
  'https://imagedelivery.net/DYKOWp0iCc0sIkF-2e4dNw/2d38ee9a-2df1-4dd0-155b-11404014bc00/public'
export const UNKNOWN_LOGO_URL =
  'https://imagedelivery.net/DYKOWp0iCc0sIkF-2e4dNw/1fb4f29a-9aed-4349-4e0b-0db6b28e7500/public'

export const INJ_DENOM = 'inj'
export const ETH_DENOM = 'peggy0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'
export const USDT_DENOM = 'peggy0xdAC17F958D2ee523a2206206994597C13D831ec7'
export const BINANCE_DEPOSIT_ADDRESS =
  'inj1u2rajhqtptzvu23leheta9yg99k3hazf4waf43'

export const ZERO_IN_WEI: BigNumberInWei = new BigNumberInWei(0)
export const ZERO_IN_BASE: BigNumberInBase = new BigNumberInBase(0)
export const NUMBER_REGEX = new RegExp(/^-?(0|[1-9]\d*)?(\.\d+)?$/)

export const GWEI_IN_WEI: BigNumber = new BigNumber(1000000000)
export const DEFAULT_GAS_PRICE = new BigNumber(120).times(GWEI_IN_WEI)
export const DEFAULT_MAINNET_GAS_PRICE = new BigNumber(30).times(GWEI_IN_WEI)
export const INJ_REQUIRED_FOR_GAS = 0.005
