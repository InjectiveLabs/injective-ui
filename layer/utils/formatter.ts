import {
  BigNumber,
  BigNumberInWei,
  BigNumberInBase
} from '@injectivelabs/utils'
import { Coin } from '@injectivelabs/sdk-ts'
import { TimeDuration } from './../types'

export const sharedToBalanceInWei = ({
  value,
  decimalPlaces = 18
}: {
  value: string | number
  decimalPlaces?: number
}): BigNumberInBase => {
  return new BigNumberInBase(10).pow(decimalPlaces).times(value)
}

export const sharedToBalanceInTokenInBase = ({
  value,
  decimalPlaces = 18
}: {
  value: string | number
  decimalPlaces?: number
}): BigNumberInBase => {
  return new BigNumberInWei(value).toBase(decimalPlaces)
}

export const sharedToBalanceInToken = ({
  value,
  fixedDecimals,
  roundingMode,
  decimalPlaces = 18
}: {
  value: string | number
  decimalPlaces?: number
  fixedDecimals?: number
  roundingMode?: BigNumber.RoundingMode
}): string => {
  const balanceInToken = sharedToBalanceInTokenInBase({
    value,
    decimalPlaces
  })

  if (fixedDecimals) {
    return balanceInToken.toFixed(fixedDecimals, roundingMode)
  }

  return balanceInToken.toFixed()
}

export const sharedParseRouteQuery = (value: string): string =>
  value.replaceAll('-', '').toLowerCase().trim()

export const sharedFormatSecondsToDisplay = ({
  value,
  roundUp = true
}: {
  value: number
  roundUp?: boolean
}) => {
  if (value === 0) {
    return
  }

  const output = {
    [TimeDuration.Day]: Math.floor(value / (3600 * 24)),
    [TimeDuration.Hour]: Math.floor((value % (3600 * 24)) / 3600),
    [TimeDuration.Minute]: Math.floor((value % 3600) / 60),
    [TimeDuration.Second]: Math.floor(value % 60)
  } as Record<string, number>

  if (!roundUp) {
    return output
  }

  if (output[TimeDuration.Day] === 1 && output[TimeDuration.Hour] === 0) {
    output[TimeDuration.Day] = 0
    output[TimeDuration.Hour] = 24
  }

  if (output[TimeDuration.Hour] === 1 && output.minutes === 0) {
    output[TimeDuration.Hour] = 0
    output[TimeDuration.Minute] = 60
  }

  if (output[TimeDuration.Minute] === 1 && output[TimeDuration.Second] === 0) {
    output[TimeDuration.Minute] = 0
    output[TimeDuration.Second] = 60
  }

  return output
}

export const sharedConvertTimestampToMilliseconds = (
  timestamp: number | string
): number => {
  const timestampInBigNumber = new BigNumberInBase(timestamp)

  if (timestamp.toString().length > 13) {
    const formatNumberBy = new BigNumberInBase(10).pow(
      timestamp.toString().length - 13
    )

    const formattedValue = timestampInBigNumber
      .dividedBy(formatNumberBy)
      .toFixed(0, BigNumber.ROUND_HALF_UP)

    return new BigNumberInBase(formattedValue).toNumber()
  }

  if (timestamp.toString().length < 13) {
    const trailingZeros = 13 - timestamp.toString().length

    return timestampInBigNumber.times(10 ** trailingZeros).toNumber()
  }

  return timestampInBigNumber.toNumber()
}

export const sharedGetExactDecimalsFromNumber = (
  number: number | string
): number => {
  if (!number.toString().includes('.')) {
    return 0
  }

  if (Number(number) % 1 === 0) {
    return 0
  }

  const [, decimals] = number.toString().split('.')

  if (!decimals) {
    return 0
  }

  return decimals.length
}

export const sharedGetTensMultiplier = (number: number | string): number => {
  const numberToBn = new BigNumber(number)

  if (numberToBn.eq(1)) {
    return 0
  }

  if (numberToBn.lt(1)) {
    return -1 * sharedGetExactDecimalsFromNumber(numberToBn.toFixed())
  }

  const [, zerosInTheNumber] = numberToBn.toFixed().split('1')

  return zerosInTheNumber.length
}

export const sharedEllipsisFormatText = (text: string, length = 20): string => {
  return text.length > length
    ? `${text.slice(0, length)}...${text.slice(
        text.length - length,
        text.length
      )}`
    : text
}

export const sharedCoinStringToCoins = (coinString: string): Coin[] => {
  return coinString.split(',').map((coin) => {
    // is used to remove any empty strings that might result from the split operation.
    const [amount, denom] = coin.split(/(\d+)/).filter(Boolean)

    return {
      denom,
      amount
    }
  })
}
