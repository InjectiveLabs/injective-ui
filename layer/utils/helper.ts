import { sharedTokenClient, tokenStaticFactory } from '../Service'
import type { TokenStatic } from '@injectivelabs/sdk-ts'

export const sharedGetToken = async (
  denomOrSymbol: string
): Promise<undefined | TokenStatic> => {
  const token = tokenStaticFactory.toToken(denomOrSymbol)

  if (token) {
    return token
  }

  const asyncToken = await sharedTokenClient.queryToken(denomOrSymbol)

  return asyncToken
}
