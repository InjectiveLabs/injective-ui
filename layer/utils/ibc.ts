import { TokenType } from '@injectivelabs/sdk-ts'
import {
  canonicalChannelIds,
  canonicalChannelsToChainList
} from './../data/ibc'
import type { TokenStatic } from '@injectivelabs/sdk-ts'

export const isIbcTokenCanonical = (token: TokenStatic) => {
  const { denom } = token

  if (!denom.startsWith('ibc/') || !token.path) {
    return false
  }

  const pathParts = token.path.replace('transfer/', '').split('/')

  /** More than one channelId */
  if (pathParts.length > 1) {
    return false
  }

  const [channelId] = pathParts

  return canonicalChannelIds.includes(channelId)
}

export const canonicalChannelsToChainListFromInjective =
  canonicalChannelsToChainList.filter((item) => item.chainA === 'Injective')

export const getIbcDestinationChain = ({
  channelPaths,
  channel,
  index,
  token
}: {
  index: number
  channel: string
  token: TokenStatic
  channelPaths: string[]
}) => {
  if (token.tokenType !== TokenType.Ibc) {
    return ''
  }

  const foundChannelFromInjective =
    canonicalChannelsToChainListFromInjective.find(
      (item) => item.channelId === channel
    )

  if (foundChannelFromInjective) {
    return foundChannelFromInjective.chainB
  }

  if (index === channelPaths.length - 1) {
    return token.name
  }

  return ''
}

export const formatNonCanonicalIbcTokenName = (token: TokenStatic): string => {
  const channelId = token.channelId
  const baseDenom = token.baseDenom

  if (!channelId || !baseDenom) {
    return 'Unknown'
  }

  const formattedDenomTrace = (token?.channelId || '').replaceAll(
    'transfer/',
    ''
  )
  const channelToLastChain = formattedDenomTrace.split('/').shift()

  const foundChannelFromInjective =
    canonicalChannelsToChainListFromInjective.find(
      (item) => item.channelId === channelToLastChain
    )

  const lastChain = foundChannelFromInjective
    ? foundChannelFromInjective.chainB
    : 'Unknown'

  // Fix for very long base denoms
  if (baseDenom.startsWith('cw20:') || baseDenom.startsWith('factory')) {
    return `${token.symbol.toUpperCase()}-${lastChain.toLowerCase()}-${channelToLastChain}`
  }

  return `${baseDenom.toUpperCase()}-${lastChain.toLowerCase()}-${channelToLastChain}`
}
