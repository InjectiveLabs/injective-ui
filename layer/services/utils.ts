import { DEFAULT_GAS_PRICE } from '../utils/constant'
import type { TransactionOptions } from '@injectivelabs/ts-types'

export const TX_DEFAULTS_GAS = 80_000_000
export const ALLOWANCE_DEFAULT_GAS_LIMIT = 45000
export const PEGGY_TRANSFER_DEFAULT_GAS_LIMIT = 100000

export const getTransactionOptions = (
  transactionOptions: Partial<TransactionOptions>
): TransactionOptions => ({
  from: transactionOptions.from,
  gas: transactionOptions.gas ? transactionOptions.gas : TX_DEFAULTS_GAS,
  gasPrice: transactionOptions.gasPrice
    ? transactionOptions.gasPrice.toString()
    : DEFAULT_GAS_PRICE.toString()
})

export const peggyDenomToContractAddress = (denom: string): string =>
  denom.replace('peggy', '')

export const getKeyFromRpcUrl = (rpcUrl: string) => {
  if (!rpcUrl.includes('alchemyapi.io') && !rpcUrl.includes('alchemy.com')) {
    return rpcUrl
  }

  const [key] = rpcUrl.split('/').reverse()

  return key
}
