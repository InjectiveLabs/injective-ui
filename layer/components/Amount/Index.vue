<script setup lang="ts">
import { BigNumberInBase } from "@injectivelabs/utils";
import { DEFAULT_ABBREVIATION_THRESHOLD } from "../../utils/constant";

const props = withDefaults(
  defineProps<{
    decimals?: number;
    useSubscript?: boolean;
    noTrailingZeros?: boolean;
    shouldAbbreviate?: boolean;
    abbreviationThreshold?: number;
    amount: string | number | BigNumberInBase;
  }>(),
  {
    decimals: 8,
    noTrailingZeros: true,
    shouldAbbreviate: true,
    abbreviationThreshold: DEFAULT_ABBREVIATION_THRESHOLD,
  },
);

const decimals = computed(() => {
  if (
    !!props.abbreviationThreshold &&
    new BigNumberInBase(props.amount || 0).gt(props.abbreviationThreshold)
  ) {
    return 2;
  }

  return props.decimals;
});
</script>

<template>
  <SharedAmountBase
    v-bind="{
      amount,
      decimals,
      useSubscript,
      noTrailingZeros,
      shouldAbbreviate,
      abbreviationThreshold,
    }"
  >
    <template #prefix>
      <slot name="prefix" />
    </template>
  </SharedAmountBase>
</template>
