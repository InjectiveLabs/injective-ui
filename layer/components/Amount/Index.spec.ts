import { mountSuspended } from '@nuxt/test-utils/runtime'
import { BigNumber } from '@injectivelabs/utils'
import { describe, expect, test } from 'vitest'
import Index from './Index.vue'

describe('Amount/Index.vue', () => {
  describe('default behavior', () => {
    const testCases = [
      { input: '0.01100000', output: '0.011' },
      { input: '1234.12345678', output: '1,234.123456' },
      { input: '1.01', output: '1.01' },
      { input: '1.1', output: '1.1' },
      { input: '1000000000.123', output: '≈1B' },
      { input: '1234.1200000005253', output: '1,234.12' },
      { input: '1.00030004', output: '1.0003' },
      { input: '1234.1234565253', output: '1,234.123456' },
      { input: '1234567', output: '≈1.2M' }
    ]

    test.each(testCases)(
      'formats $input to $output by default',
      async ({ input, output }) => {
        const component = await mountSuspended(Index, {
          props: {
            amount: input
          }
        })

        expect(component.text()).toBe(output)
      }
    )

    describe('small number subscripts', () => {
      const smallNumberCases = [
        '0.0000001',
        '-0.0000001',
        '0.000000000001',
        '0.000000000001234',
        '1.000000000001234'
      ]

      test.each(smallNumberCases)(
        'formats small number %s with subscripts',
        async (amount) => {
          const component = await mountSuspended(Index, {
            props: {
              amount
            }
          })
          expect(component.html()).toMatchSnapshot()
        }
      )
    })
  })

  describe('full value mode (no abbreviation)', () => {
    const props = {
      shouldAbbreviate: false,
      showSmallerThan: false
    }

    describe('with subscripts enabled', () => {
      const testCases = [
        { input: '0.0000001', expected: '0.0<sub>6</sub>1' },
        { input: '1000000000.123', expected: '1,000,000,000.123' },
        { input: '1234567', expected: '1,234,567' },
        { input: '0.000000000001234', expected: '0.0<sub>11</sub>1234' },
        { input: '1.000000000001234', expected: '1' } // This number is >= 1, so no subscript
      ]

      test.each(testCases)(
        'formats $input correctly with subscripts',
        async ({ input, expected }) => {
          const component = await mountSuspended(Index, {
            props: {
              amount: input,
              useSubscript: true,
              ...props
            }
          })

          expect(component.html()).toContain(expected)
        }
      )
    })

    describe('with subscripts disabled', () => {
      const testCases = [
        { input: '0.0000001', expected: '0' },
        { input: '1000000000.123', expected: '1,000,000,000.123' },
        { input: '1234567', expected: '1,234,567' },
        { input: '0.000000000001234', expected: '0' },
        { input: '1.000000000001234', expected: '1' }
      ]

      test.each(testCases)(
        'formats $input correctly without subscripts',
        async ({ input, expected }) => {
          const component = await mountSuspended(Index, {
            props: {
              amount: input,
              useSubscript: false,
              ...props
            }
          })

          expect(component.text()).toBe(expected)
        }
      )
    })
  })

  describe('abbreviation behavior', () => {
    test('abbreviates large numbers by default', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '2000001'
        }
      })

      expect(component.text()).toBe('≈2M')
    })

    test('does not abbreviate when shouldAbbreviate is false', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '2000000',
          shouldAbbreviate: false
        }
      })

      expect(component.text()).toBe('2,000,000')
    })

    test('respects custom abbreviation threshold', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '500001',
          abbreviationThreshold: 100000
        }
      })

      expect(component.text()).toBe('≈500K')
    })
  })

  describe('smaller than behavior', () => {
    test('shows "smaller than" for very small numbers by default', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '0.0000001'
        }
      })

      expect(component.html()).toContain('&lt;0.000001')
    })

    test('shows actual value when showSmallerThan is false', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '0.0000001',
          showSmallerThan: false
        }
      })

      expect(component.html()).toContain('0.0<sub>6</sub>1')
    })
  })

  describe('decimal precision', () => {
    test('respects custom decimals', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1.123456789',
          decimals: 2
        }
      })

      expect(component.text()).toBe('1.12')
    })

    test('adjusts decimals for abbreviated numbers', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1500001',
          decimals: 6
        }
      })

      expect(component.text()).toBe('≈1.5M')
    })
  })

  describe('trailing zeros', () => {
    test('removes trailing zeros by default', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1.10000'
        }
      })

      expect(component.text()).toBe('1.1')
    })

    test('keeps trailing zeros when noTrailingZeros is false', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1.10000',
          noTrailingZeros: false
        }
      })

      expect(component.text()).toBe('1.100000')
    })
  })

  describe('zero handling', () => {
    test('shows zero normally by default', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '0'
        }
      })

      expect(component.text()).toBe('0')
    })

    test('shows em-dash for zero when showZeroAsEmDash is true', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '0',
          showZeroAsEmDash: true
        }
      })

      expect(component.html()).toContain('—')
    })
  })

  describe('negative numbers', () => {
    test('handles negative numbers correctly', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '-1234.567'
        }
      })

      expect(component.text()).toBe('-1,234.567')
    })

    test('handles negative small numbers with subscripts', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '-0.0000123'
        }
      })

      expect(component.html()).toMatchInlineSnapshot(
        `"<span><span>-</span><span>0.0<sub>4</sub>123</span></span>"`
      )
    })

    test('handles negative numbers in full value mode', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '-0.0000123',
          shouldAbbreviate: false,
          showSmallerThan: false
        }
      })

      expect(component.html()).toMatchInlineSnapshot(
        `"<span><span>-</span><span>0.0<sub>4</sub>123</span></span>"`
      )
    })
  })

  describe('rounding modes', () => {
    test('uses ROUND_DOWN by default', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1.999999',
          decimals: 2
        }
      })

      expect(component.text()).toBe('1.99')
    })

    test('respects custom rounding mode', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1.999999',
          decimals: 2,
          roundingMode: BigNumber.ROUND_UP,
          noTrailingZeros: false
        }
      })

      expect(component.text()).toBe('2.00')
    })
  })

  describe('prop combinations', () => {
    test('full value mode with custom decimals', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1000000.123456789',
          decimals: 4,
          shouldAbbreviate: false,
          showSmallerThan: false
        }
      })

      expect(component.text()).toBe('1,000,000.1234')
    })

    test('subscripts with custom threshold', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '0.001234',
          useSubscript: true,
          showSmallerThan: false
        }
      })

      expect(component.html()).toContain('0.001234')
    })

    test('no trailing zeros with custom decimals', async () => {
      const component = await mountSuspended(Index, {
        props: {
          amount: '1.5',
          decimals: 4,
          noTrailingZeros: true
        }
      })

      expect(component.text()).toBe('1.5')
    })
  })
})
