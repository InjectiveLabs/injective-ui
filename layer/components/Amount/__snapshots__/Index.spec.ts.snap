// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Amount/Index.vue > default behavior > small number subscripts > formats small number -0.0000001 with subscripts 1`] = `"<span><span>-</span><span>&lt;0.000001</span></span>"`;

exports[`Amount/Index.vue > default behavior > small number subscripts > formats small number 0.000000000001 with subscripts 1`] = `"<span><!--v-if--><span>&lt;0.000001</span></span>"`;

exports[`Amount/Index.vue > default behavior > small number subscripts > formats small number 0.000000000001234 with subscripts 1`] = `"<span><!--v-if--><span>&lt;0.000001</span></span>"`;

exports[`Amount/Index.vue > default behavior > small number subscripts > formats small number 0.0000001 with subscripts 1`] = `"<span><!--v-if--><span>&lt;0.000001</span></span>"`;

exports[`Amount/Index.vue > default behavior > small number subscripts > formats small number 1.000000000001234 with subscripts 1`] = `"<span><!--v-if--><span>1</span></span>"`;
